[AMDGPU] Automatic conversion from wave32 to wave64

alex-t opened the pull request:
Small short living kernels may become waveslot limited. To work around the problem an optimization is proposed to convert such kernels from wave32 to wave64 automatically. These kernels shall conform to a strict set of limitations and satisfy profitability conditions.

1. A kernel shall have no function calls as we cannot analyze call stack requirements (nor will it fall into a category of short living kernels anyway).
2. A kernel itself shall not be called from a device enqueue call.
3. A kernel shall not attempt to access EXEC or VCC in any user visible way.
4. A kernel must not use readlane/readfirstlane or any cross-lane/DPP operations in general.
5. A kernel shall not read wavefront size or use ballot through intrinsics (a use of pre-defined frontend wave size macro was deemed permissible for now).
6. There shall be no atomic operations of any sort as these may be used for cross-thread communication.
7. There shall be no LDS access as the allocation is usually tied to the workgroup size and we generally cannot extend it. It is also changing occupancy which is tied to the wave size.
8. There shall be no inline asm calls.
9. There shall be no dynamic VGPRs.
10. Starting from GFX11 some instructions (such as WMMA on GFX11+ and transpose loads on GFX12+) work differently (have different operands) in wave32 and wave64. The kernel shall not have intrinsics to invoke such instructions.

Commit:
[AMDGPU] Automatic conversion from wave32 to wave64

alex-t commented:
This is an early draft and was created as a discussion board to elaborate the proper approach and solutions.

alex-t commented:
A few questions to discuss right now:
1. May we assume that for the targets which have a wave32 as default correct attribute will be set by the FE?
2. Non-convergent intrinsic calls are left in the IR as a calls. I guess we should change the callee version (.i32/i64) if we set wave64 on the caller.

rampitec commented on llvm/lib/Target/AMDGPU/SIConvertWaveSize.cpp:
```diff
@@ -0,0 +1,321 @@
+        }
+      }
+      // No  LDS access is allowed
+      if (auto LI = dyn_cast<LoadInst>(&I)) {
```
Load or store to local address is not a sufficient check. LDS may be addressed by flat. The kernel shall have no references to LDS objects.

alex-t replied:
I have a few questions...

1. We can pass a flat pointer as an argument. 

- We cannot check that it does not point to LDS. 
- "noalias" attribute does not help. 
- AMDGPU::addrspacesMayAlias states that the Flat may alias to anything. - **On the other hand, it would be ABI violation if kernel argument aliased the object in LDS.** 

3. We can have a non-pointer (i32) argument and int2ptr to addrspace 3 - trivial to check
4. addrspacecast to addrspace 3 - trivial to check
5. Any arbitrary integer expression cast to flat pointer. - We cannot check if it won't point to LDS (probably corner case)

rampitec replied:
You cannot pass a pointer to LDS into a kernel. There is no way to allocate it. But dynamic LDS shall be disallowed.

arsenm replied:
But a kernel can have an LDS allocation as a kernel argument, as how it works in opencl. It has no content

rampitec replied:
Right, that's what I mean. Transformation shall be disabled if we have dynamic LDS.

alex-t replied:
I already added a few checks for dynamic LDS such as:

- kernel arguments of addrspace 3
- addrspacecast/bitcast/int2ptr with destination pointer addrspace 3 (assuming the flat 2 LDS casts).

what else could you suggest?

rampitec replied:
A kernel cannot have LDS pointer arguments because it is impossible to allocate LDS on the host. There is a way for host to *ask* LDS to be allocated at wave launch, which is dynamic LDS.

alex-t replied:
> A kernel cannot have LDS pointer arguments because it is impossible to allocate LDS on the host. There is a way for host to _ask_ LDS to be allocated at wave launch, which is dynamic LDS.

Have no idea how I can check this in IR

alex-t replied:
So, this: `define amdgpu_kernel void @test_lds_access(ptr addrspace(3) %out) ` is syntactically correct but impossible, right?

rampitec replied:
Yes, it is impossible.

arsenm replied:
No, that is how dynamic LDS is represented in OpenCL

alex-t replied:
> But a kernel can have an LDS allocation as a kernel argument, as how it works in opencl. It has no content

How it looks like? Like a flat pointer that is assigned a value on runtime to point to LDS? If so, the "addrspacecast/inttoptr" check will help to detect this...

arsenm replied:
Just as you wrote, a `ptr addrspace(3)` kernel argument

rampitec commented on llvm/lib/Target/AMDGPU/SIConvertWaveSize.cpp:
```diff
@@ -0,0 +1,321 @@
+
+      // TODO: Dynamic VGPRS and GFX11+ special operations ???
+      BlockCost +=
+          TTI->getInstructionCost(&I, TargetTransformInfo::TCK_RecipThroughput);
```
TCK_Latency maybe?

rampitec commented on llvm/lib/Target/AMDGPU/SIConvertWaveSize.cpp:
```diff
@@ -0,0 +1,321 @@
+  }
+
+bool SIConvertWaveSize::changeWaveSizeAttr(Function *F) {
+  auto Attr = F->getFnAttribute("target-features");
```
In addition you need to change AMDGPUSubtarget::WavefrontSizeLog2. ST.getWavefrontSize() should return correct value.

rampitec replied:
I mean you need to change this variable.

alex-t replied:
> I mean you need to change this variable.

I still don't understand.
This "variable" is a protected member of a AMDGPUSubtarget class, to change it I need to change the class interface.

rampitec replied:
If you do not change it the whole BE will think you are in wave32 mode. If you need to change the interface, so be it.

rampitec commented on llvm/test/CodeGen/AMDGPU/wave32-to-64-auto-convert.ll:
```diff
@@ -0,0 +1,121 @@
+; RUN: opt -S -mtriple=amdgcn -mcpu=gfx1100 -passes=si-convert-wave-size < %s | FileCheck %s
```
Also need to test full compilation and check resulting .amdhsa_wavefront_size32 value.

arsenm commented on llvm/lib/Target/AMDGPU/SIConvertWaveSize.cpp:
```diff
@@ -0,0 +1,321 @@
+//===- SIConvertWaveSize.cpp - Automatically converts wave32 kernels to wave64
+//---------===//
```
Bad line break

arsenm commented on llvm/lib/Target/AMDGPU/SIConvertWaveSize.cpp:
```diff
@@ -0,0 +1,321 @@
+// kernels from wave32 to wave64 automatically.These kernels shall conform to a
+// strict set of limitations and satisfy profitability conditions.
+//
+// 1. A kernel shall have no function calls as we cannot analyze call stack
+// requirements (nor will it fall into a category of short living kernels
+// anyway).
```
If you did this in the attributor and could convert the entire call graph this wouldn't be an issue

rampitec replied:
Not needed though. A cost of one call will make the whole transformation useless.

arsenm replied:
The call itself isn't necessarily expensive, but the attributor would also take care of most of these queries even within a single kernel

rampitec replied:
We are talking about quite low latency threshold.

arsenm commented on llvm/lib/Target/AMDGPU/SIConvertWaveSize.cpp:
```diff
@@ -0,0 +1,321 @@
+
+namespace {
+class SIConvertWaveSize {
+  const TargetMachine *TM;
```
Use GCNTargetMachine

arsenm commented on llvm/lib/Target/AMDGPU/SIConvertWaveSize.cpp:
```diff
@@ -0,0 +1,321 @@
+}
+
+bool SIConvertWaveSize::run(Function &F) {
+  LLVM_DEBUG(dbgs() << "Running SIConvertWaveSize on function: " << F.getName() << "\n");
+  LLVM_DEBUG(printFunctionAttributes(F));
```
Remove this, pass manager should be responsible for this type of information

arsenm commented on llvm/lib/Target/AMDGPU/SIConvertWaveSize.cpp:
```diff
@@ -0,0 +1,321 @@
+};
+} // end anonymous namespace
+
+void printFunctionAttributes(const Function &F) {
+  LLVM_DEBUG(dbgs() << "Function: " << F.getName() << "\n");
+  for (const auto &Attr : F.getAttributes()) {
+    LLVM_DEBUG(dbgs() << "  Attribute: " << Attr.getAsString() << "\n");
+  }
+}
```
Leftover debug

alex-t replied:
It was left for the purpose as this is an early draft. I will remove it once the debugging is finished.

arsenm commented on llvm/lib/Target/AMDGPU/SIConvertWaveSize.cpp:
```diff
@@ -0,0 +1,321 @@
+    return false;
+
+  // Check if the kernel is wave32
+  if (F.hasFnAttribute("target-features")) {
```
Don't try to parse subtarget features manually, directly read from GCNSubtarget

arsenm commented on llvm/lib/Target/AMDGPU/SIConvertWaveSize.cpp:
```diff
@@ -0,0 +1,321 @@
+  }
+
+  // Check if the function is a device enqueue call.
+  if (F.hasFnAttribute("amdgpu-device-enqueue")) {
```
We don't have any more of these broken positive enable features. We infer no use of the device enqueue inputs. In this context, probably should just make sure the address of the function isn't captured

arsenm commented on llvm/lib/Target/AMDGPU/SIConvertWaveSize.cpp:
```diff
@@ -0,0 +1,321 @@
+    return false;
+  }
+
+  // Check if a trip count is a compile time constant for all loops in the
+  // kernel
+  for (Loop *L : *LI) {
+    const SCEV *TripCountSCEV = SE->getBackedgeTakenCount(L);
+    if (!isa<SCEVConstant>(TripCountSCEV)) {
+      LLVM_DEBUG(
+          dbgs() << "SIConvertWaveSize: Trip count is not a compile time "
+                    "constant.\n");
+      return false;
+    }
+  }
```
I don't understand why this is a necessary condition. What if there are unstructured cycles, they won't be detected?

arsenm commented on llvm/lib/Target/AMDGPU/SIConvertWaveSize.cpp:
```diff
@@ -0,0 +1,321 @@
+                     << "SIConvertWaveSize: Atomic operation detected.\n");
+          return false;
+        }
+        if (Function *Callee = CB->getCalledFunction()) {
```
Should directly check if the callsite is convergent, without looking at the callee directly

arsenm commented on llvm/lib/Target/AMDGPU/SIConvertWaveSize.cpp:
```diff
@@ -0,0 +1,321 @@
+                }
+              }
+
+            if (Callee->getIntrinsicID() == Intrinsic::read_register) {
```
Should express this in terms of attributes. Probably need nosync + nocallback. Not sure we can get away without listing read_register/write_register, but maybe we can

alex-t replied:
Also, I would like to elaborate the "whitelist" of the convergent intrinsics. What else except for "amdgcn_s_barrier"?

rampitec replied:
_now_ this is just s_barrier. It can be extended in the future.

alex-t replied:
>You can just bail on any use of these, no reason to analyze what really does it touch.
We could convert more kernels if we allow read/write_register if they don't touch EXEC or VCC. Given that we change all uses of read/write_register.i32 to *.i64 to ensure they read the full register, not its lower part. Does it make sense?

arsenm replied:
These intrinsics don't get used in the real world, don't worry about them and just disallow them. We used to use them for exec for ballot, but switched to just using ballot

rampitec replied:
> > You can just bail on any use of these, no reason to analyze what really does it touch.
> 
> We could convert more kernels if we allow read/write_register if they don't touch EXEC or VCC. Given that we change all uses of read/write_register.i32 to *.i64 to ensure they read the full register, not its lower part. Does it make sense?

Just probibit these completely.

arsenm commented on llvm/lib/Target/AMDGPU/SIConvertWaveSize.cpp:
```diff
@@ -0,0 +1,321 @@
+          LLVM_DEBUG(dbgs() << "SIConvertWaveSize: LDS access detected.\n");
+          return false;
+        }
+      }
```
Also misses all memory intrinsics and atomics

arsenm commented on llvm/lib/Target/AMDGPU/SIConvertWaveSize.cpp:
```diff
@@ -0,0 +1,321 @@
+                     << "SIConvertWaveSize: Inline assembly detected.\n");
+          return false;
+        }
+        if (CB->isAtomic()) {
```
Don't think this check does anything useful on an arbitrary call, it's a wrapper around load/store/atomicrmw/cmpxchg

arsenm commented on llvm/lib/Target/AMDGPU/SIConvertWaveSize.cpp:
```diff
@@ -0,0 +1,321 @@
+  return Changed;
+  }
+
+bool SIConvertWaveSize::changeWaveSizeAttr(Function *F) {
```
You shouldn't be trying to parse or find the position, you can simply append the new wavesize feature at the end

arsenm commented on llvm/test/CodeGen/AMDGPU/wave32-to-64-auto-convert.ll:
```diff
@@ -0,0 +1,121 @@
+bb4:
+  br i1 %tmp3, label %bb1, label %bb2
+}
+
```
test atomicrmw + cmpxchg + memory intrinsics

arsenm reviewed (commented):
This would be easier to implement in attributor

arsenm commented on llvm/lib/Target/AMDGPU/SIConvertWaveSize.cpp:
```diff
@@ -0,0 +1,321 @@
+      std::string End = AttrStr.substr(Pos + strlen("+wavefrontsize32")).str();
+      std::string NewAttrStr = NewBegin + End;
+      // Add the "+wavefrontsize64" attribute.
+      F->removeFnAttr("target-features");
+      F->addFnAttr("target-features", NewAttrStr);
```
just doing addFnAttr should override the old value

arsenm reviewed (commented):
This also probably should be an opt-in feature... I don't think we have any API rules that prevent you from assuming the wavesize will match what you compiled for. At the very least we need an explicit opt-out mechanism

alex-t commented on llvm/lib/Target/AMDGPU/SIConvertWaveSize.cpp:
```diff
@@ -0,0 +1,321 @@
+  }
+
+  // Check if the function is a device enqueue call.
+  if (F.hasFnAttribute("amdgpu-device-enqueue")) {
```
Just forgot to comment it out.  It does not exist yet.  The placeholder was added to mark an idea to discuss: could we make the FE to set such an attribute basing on the @llvm.amdgcn.dispatch.ptr, @llvm.amdgcn.implicitarg.ptr and other relevant attributes presence?

rampitec replied:
I believe kernel just shall have no references to its symbol, i.t. its address shall not escape.

arsenm replied:
No, we can only do negative usage detection in AMDGPUAttributor. Attributes that specify a feature are positively used are broken, we don't have any of those anymore.

alex-t commented on llvm/lib/Target/AMDGPU/SIConvertWaveSize.cpp:
```diff
@@ -0,0 +1,321 @@
+    return false;
+
+  // Check if the kernel is wave32
+  if (F.hasFnAttribute("target-features")) {
```
What if the "target-features" was set manually in the IR and does not match to the command line option (i.e. subtarget feature)? Currently it leads to the "target-features"="+wavefrontsize32,+wavefrontsize64" appearance, that apparently is a bug.

rampitec replied:
This just should not happen. You cannot pass -mattr to clang and llc shall not be used for production. If you see that happens we should emit error.

arsenm replied:
If target-features is explicit in the IR, it is directly used. If it is not set, it comes from the TargetMachine's default target features. In the case it's unset, you need to pull the default from the TM and append to that

Commit:
[AMDGPU] Automatic conversion from wave32 to wave64. Review issues addressed.

Commit:
[AMDGPU] Automatic conversion from wave32 to wave64. Initialize method name fixed

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUConvertWaveSize.cpp:
```diff
@@ -0,0 +1,403 @@
+    }
+  }
+
+  // Check if the function can be called via device enqueue.
+  bool addressEscapes = false;
+  if (!F.use_empty()) {
+    const Module *M = F.getParent();
+    for (const GlobalVariable &GV : M->globals()) {
+      if (GV.hasInitializer()) {
+        if (const Constant *Init = GV.getInitializer()) {
+          if (isa<Function>(Init) && Init == &F) {
+            addressEscapes = true;
+          }
+        }
+      }
+    }
+  }
+
+  if (addressEscapes) {
```
F.hasAddressTaken (but again, this whole thing would be easier to implement in attributor)

shiltian replied:
It would be overkill, since this pass gives up if there is any function call so there is no inter-procedural analysis required.

alex-t commented:
The last commit is technical - just to keep things in order. Addresses some comments. The LIT test checks are not functional anymore. Most likely opt run will be replaced with the full compilation so that we can check the wavefrontsize for each kernel.

Commit:
[AMDGPU] Automatic conversion from wave32 to wave64.
  -- Simplified wavefrontsize attribute update
  -- LDS access checks changed
  -- new LDS test in LIT NOTE: because of the simplified attribute
update need to change all the checks. Most likely the direct checks of
attributes is now impossible.

alex-t commented on llvm/lib/Target/AMDGPU/AMDGPU.h:
```diff
@@ -174,6 +177,9 @@ extern char &SIShrinkInstructionsLegacyID;
 extern char &SIFixSGPRCopiesLegacyID;
 
+void initializeAMDGPUConvertWaveSizeLegacyPass(PassRegistry &);
+extern char &AMDGPUConvertWaveSizeLegacyID;
```
The opt pipeline for sure as it works on the IR - not MIR. The method above was added for Legacy PM.
